### PR TITLE
Auth impl

### PR DESCRIPTION
**Pull request by manojkrsaini09**
Added spring security with user, role table. Added client side handling and a simple login page.

### CODE REVIEW

File: src/main/java/creativei/configuration/Filters/InitializationFilter.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+@Component
+@Order(1)
+public class InitializationFilter implements Filter {
+    private static final Logger logger = LoggerFactory.getLogger(InitializationFilter.class);
+
+    @Autowired
+    UserService userService;
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/Filters/InitializationFilter.java]: Do we need to implement blank definaiton?

File: src/main/java/creativei/configuration/Filters/InitializationFilter.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                logger.error(e.getMessage(), e);
+            }
+        }
+
+        filterChain.doFilter(servletRequest,servletResponse);
+    }
+
+    @Override
+    public void destroy() {
+
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/Filters/InitializationFilter.java]: Add log statement.

File: src/main/java/creativei/configuration/WebSecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  @Autowired
+  private CustomLogoutSuccessfulHandler logoutSuccessfulHandler;
+
+  @Autowired
+  private CustomAccessDeniedHandler customAccessDeniedHandler;
+
+  @Autowired
+  private CustomAuthenticationEntryPoint customAuthenticationEntryPoint;
+
+  @Resource(name = "customeUserDetailService")
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/WebSecurityConfig.java]: Spell check?

File: src/main/java/creativei/configuration/WebSecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.IOException;
+
+@Configuration
+@EnableWebSecurity
+@EnableGlobalMethodSecurity(
+        prePostEnabled = true, securedEnabled = true, jsr250Enabled = true
+)
+public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
+  private static final Logger logger = LoggerFactory.getLogger(WebSecurityConfig.class);
+  private static final ObjectMapper objectMapper = new ObjectMapper();
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/WebSecurityConfig.java]: Can we use autowiring?

File: src/main/java/creativei/configuration/WebSecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+              .antMatchers("/index.html", "/", "/login" ,"/userInfo",
+                      "/**.js","/module/**","/assets/**","/languages/**","/createUser",
+                      "/logout",  "/**.png","/**.jpeg","/resources/**","/assets/**").permitAll()
+              .anyRequest().authenticated()
+              .and().exceptionHandling().accessDeniedHandler(customAccessDeniedHandler)
+              .authenticationEntryPoint(customAuthenticationEntryPoint).and()
+              .logout().deleteCookies("JSESSIONID")
+              .and()
+             /* .rememberMe().rememberMeParameter("remember-me").key("uniqueAndSecret")
+              .tokenValiditySeconds(600)
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/WebSecurityConfig.java]: use constants for hard-coded value.

File: src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+@Component
+public class CustomAccessDeniedHandler implements AccessDeniedHandler {
+
+  @Override
+  public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException)
+      throws IOException, ServletException {
+    response.setStatus(HttpStatus.FORBIDDEN.value());
+    response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+    response.setCharacterEncoding("UTF-8");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java]: use constants

File: src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    JSONObject jsonResponse = new JSONObject();
+    jsonResponse.put("message", "Access Denied");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java]: use constants

File: src/main/java/creativei/configuration/securityhandler/CustomAuthenticationEntryPoint.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class CustomAuthenticationEntryPoint implements AuthenticationEntryPoint {
+
+
+  @Override
+  public void commence(HttpServletRequest request, HttpServletResponse response, AuthenticationException authException)
+      throws IOException, ServletException {
+    //response.sendError(HttpStatus.UNAUTHORIZED.value(), HttpStatus.FORBIDDEN.toString());
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/securityhandler/CustomAuthenticationEntryPoint.java]: Where are we sending this error back?

File: src/main/java/creativei/configuration/securityhandler/CustomLoginFailureHandler.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Override
+  public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response, AuthenticationException exception)
+      throws IOException, ServletException {
+    
+    response.setStatus(HttpStatus.BAD_REQUEST.value());
+    response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+    response.setCharacterEncoding("UTF-8");
+    JSONObject jsonResponse = new JSONObject();
+    jsonResponse.put("message", "Invalid Credentials");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/securityhandler/CustomLoginFailureHandler.java]: use constants

File: src/main/java/creativei/entity/User.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	public String getPassword() {
+		return password;
+	}
+
+	public void setPassword(String password) {
```
Comment (aayushKhandpur) [inline src/main/java/creativei/entity/User.java]: Avoid adding the setter for Password. Use a constructor for the setting of the password. This is a security best-practice.

File: src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	@Autowired
+	private UserService userService;
+	
+	@Override
+	public UserDetails loadUserByUsername(String userName) throws UsernameNotFoundException {
+		User user;
+		try {
+			user = userService.getByUserName(userName);
+			if(user == null) {
+				logger.error("User with %s doesn't exist!", userName);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java]: Use constant and avoid duplicating comment.

File: src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+				throw new UsernameNotFoundException(String.format("User with %s doesn't exist!", userName));
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java]: Use constant and avoid duplicating comment.

File: src/main/java/creativei/configuration/Filters/InitializationFilter.java (hunk 1 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/Filters/InitializationFilter.java]: it is not blank. I am setting user information in session if it is not present. 

  @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
            throws IOException, ServletException {

        HttpServletRequest req = (HttpServletRequest) servletRequest;
        HttpServletResponse res = (HttpServletResponse) servletResponse;
        Principal principal = req.getUserPrincipal();
        if(req.getSession().getAttribute(CommonConstant.APP_USER) == null && principal!=null){
            User user = null;
            try {
                user = userService.getByUserName(principal.getName());
                req.getSession().setAttribute(CommonConstant.APP_USER,new UserVO(user));
            } catch (ApplicationLevelException e) {
                logger.error(e.getMessage(), e);
            }
        }

        filterChain.doFilter(servletRequest,servletResponse);
    }

File: src/main/java/creativei/configuration/WebSecurityConfig.java (hunk 3 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/WebSecurityConfig.java]: done

File: src/main/java/creativei/configuration/WebSecurityConfig.java (hunk 4 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/WebSecurityConfig.java]: Autowiring is used. Do you talked about @Resource? It is same as @autowiring with qualifier.

File: src/main/java/creativei/configuration/WebSecurityConfig.java (hunk 5 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/WebSecurityConfig.java]: done

File: src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java (hunk 7 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java]: done

File: src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java (hunk 6 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/securityhandler/CustomAccessDeniedHandler.java]: done

File: src/main/java/creativei/configuration/securityhandler/CustomAuthenticationEntryPoint.java (hunk 8 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/securityhandler/CustomAuthenticationEntryPoint.java]: Each access denied request passed by this interceptor. At client side then we can use response status for required action in this case.

File: src/main/java/creativei/configuration/securityhandler/CustomLoginFailureHandler.java (hunk 9 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/securityhandler/CustomLoginFailureHandler.java]: done

File: src/main/java/creativei/configuration/securityhandler/CustomLoginFailureHandler.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class CustomLoginFailureHandler implements AuthenticationFailureHandler {
+
+  @Override
+  public void onAuthenticationFailure(HttpServletRequest request, HttpServletResponse response, AuthenticationException exception)
+      throws IOException, ServletException {
+    
+    response.setStatus(HttpStatus.BAD_REQUEST.value());
```
[the remark this answers is not in the document]
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/securityhandler/CustomLoginFailureHandler.java]: done

File: src/main/java/creativei/configuration/securityhandler/CustomLoginSuccessfulHandler.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+@Component
+public class CustomLoginSuccessfulHandler implements AuthenticationSuccessHandler {
+
+  @Override
+  public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, Authentication authentication)
+      throws IOException, ServletException {
+
+    response.setStatus(HttpStatus.OK.value());
+    response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+    response.setContentType("application/json");
```
[the remark this answers is not in the document]
Comment (manojkrsaini09) [reply src/main/java/creativei/configuration/securityhandler/CustomLoginSuccessfulHandler.java]: done

File: src/main/java/creativei/entity/User.java (hunk 10 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/entity/User.java]: done

File: src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java (hunk 11 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java]: done

File: src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java (hunk 12 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/service/securityhelpers/CustomeUserDetailService.java]: done

File: src/main/java/creativei/configuration/securityhandler/CustomLoginSuccessfulHandler.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
+package creativei.configuration.securityhandler;
+
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.smartcardio.CommandAPDU;
```
Comment (aayushKhandpur) [inline src/main/java/creativei/configuration/securityhandler/CustomLoginSuccessfulHandler.java]: Why do we need this import?

**Comment by manojkrsaini09** [comment]
done changes

**Pull request merged**